internal/adaptorsigs/adaptor_test.go

```diff
@@ -0,0 +1,131 @@
+package adaptorsigs
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4/schnorr"
+)
+
+func TestAdaptorSignatureRandom(t *testing.T) {
+	seed := time.Now().Unix()
+	rng := rand.New(rand.NewSource(seed))
+	defer func(t *testing.T, seed int64) {
+		if t.Failed() {
+			t.Logf("random seed: %d", seed)
+		}
+	}(t, seed)
+
+	for i := 0; i < 100; i++ {
+		// Generate two private keys
+		var pkBuf1, pkBuf2 [32]byte
+		if _, err := rng.Read(pkBuf1[:]); err != nil {
+			t.Fatalf("failed to read random private key: %v", err)
+		}
+		if _, err := rng.Read(pkBuf2[:]); err != nil {
+			t.Fatalf("failed to read random private key: %v", err)
+		}
+		var privKey1Scalar, privKey2Scalar secp256k1.ModNScalar
+		privKey1Scalar.SetBytes(&pkBuf1)
+		privKey2Scalar.SetBytes(&pkBuf2)
+		privKey1 := secp256k1.NewPrivateKey(&privKey1Scalar)
+		privKey2 := secp256k1.NewPrivateKey(&privKey2Scalar)
+
+		// Generate random hashes to sign.
+		var hash1, hash2 [32]byte
+		if _, err := rng.Read(hash1[:]); err != nil {
+			t.Fatalf("failed to read random hash: %v", err)
+		}
+		if _, err := rng.Read(hash2[:]); err != nil {
+			t.Fatalf("failed to read random hash: %v", err)
+		}
+
+		// Generate random signature tweak
+		var tBuf [32]byte
+		if _, err := rng.Read(tBuf[:]); err != nil {
+			t.Fatalf("failed to read random private key: %v", err)
+		}
+		var tweak secp256k1.ModNScalar
+		tweak.SetBytes(&tBuf)
+
+		// Sign hash1 with private key 1
+		sig, err := schnorr.Sign(privKey1, hash1[:])
+		if err != nil {
+			t.Fatalf("Sign error: %v", err)
+		}
+
+		// The owner of priv key 1 knows the tweak. Sends a priv key tweaked adaptor sig
+		// to the owner of priv key 2.
+		adaptorSigPrivKeyTweak := PrivateKeyTweakedAdaptorSig(sig, privKey1.PubKey(), &tweak)
+		err = adaptorSigPrivKeyTweak.Verify(hash1[:], privKey1.PubKey())
+		if err != nil {
+			t.Fatalf("verify error: %v", err)
+		}
+
+		// The owner of privKey2 creates a public key tweaked adaptor sig using
+		// tweak * G, and sends it to the owner of privKey1.
+		adaptorSigPubKeyTweak, err := PublicKeyTweakedAdaptorSig(privKey2, hash2[:], adaptorSigPrivKeyTweak.PublicTweak())
+		if err != nil {
+			t.Fatalf("PublicKeyTweakedAdaptorSig error: %v", err)
+		}
+
+		// The owner of privKey1 knows the tweak, so they can decrypt the
+		// public key tweaked adaptor sig.
+		decryptedSig, err := adaptorSigPubKeyTweak.Decrypt(&tweak)
+		if err != nil {
+			t.Fatal(err)
+		}
+
```

Comment:
I'd suggest verifying the decrypted signature here to make sure it's a valid one.

```Go
		if !decryptedSig.Verify(hash2[:], privKey2.PubKey()) {
			t.Fatal("failed to verify decrypted signature")
		}
```

Reply:
It'd also ideally check all the fields (aka `s`, `t`, and `pubKeyTweak` as opposed to just `r` to catch this type of thing

Reply:
> It'd also ideally check all the fields (aka s, t, and pubKeyTweak as opposed to just r to catch this type of thing

This comment was regarding the parsing test right? I've added an `IsEqual` function to `AdaptorSignature`.

Reply:
Correct.  It was only doing `if !adaptor2.r.Equals(&adaptor.r) {` before.  I meant to comment on the actual line.

Reply:
Additional tests to make sure that private/public variants are correct for random values look good,